aws/rust-runtime/aws-config/src/imds/credentials.rs

```diff
@@ -0,0 +1,487 @@
+/// response from IMDS vs. something invalid / unexpected. The inner error distinguishes between
+/// a successful request to IMDS that includes credentials vs. an error with a code.
+fn parse_imds_credentials(
+    credentials_response: &str,
+) -> Result<ImdsCredentialsResponse, InvalidImdsResponse> {
+    let mut tokens = json_token_iter(credentials_response.as_bytes()).peekable();
+    let mut code = None;
+    let mut tpe = None;
+    let mut access_key_id = None;
+    let mut secret_access_key = None;
+    let mut session_token = None;
+    let mut expiration = None;
+    let mut message = None;
+    if let Some(Token::StartObject { .. }) = tokens.next().transpose()? {
+        loop {
+            match tokens.next().transpose()? {
+                Some(Token::EndObject { .. }) => break,
+                Some(Token::ObjectKey { key, .. }) => {
+                    if let Some(Ok(Token::ValueString { value, .. })) = tokens.peek() {
+                        match key.as_escaped_str() {
+                            /*
+                             "Code": "Success",
+                             "Type": "AWS-HMAC",
+                             "AccessKeyId" : "accessKey",
+                             "SecretAccessKey" : "secret",
+                             "Token" : "token",
+                             "Expiration" : "....",
+                             "LastUpdated" : "2009-11-23T0:00:00Z"
+                            */
+                            "Code" => code = Some(value.to_unescaped()?),
+                            "Type" => tpe = Some(value.to_unescaped()?),
+                            "AccessKeyId" => access_key_id = Some(value.to_unescaped()?),
+                            "SecretAccessKey" => secret_access_key = Some(value.to_unescaped()?),
+                            "Token" => session_token = Some(value.to_unescaped()?),
+                            "Expiration" => expiration = Some(value.to_unescaped()?),
+
+                            // Error case handling: message will be set
+                            "Message" => message = Some(value.to_unescaped()?),
+                            _ => {}
+                        }
+                    }
+                    skip_value(&mut tokens)?;
+                }
+                other => {
+                    return Err(InvalidImdsResponse::Custom(
+                        format!("expected object key, found: {:?}", other,).into(),
+                    ))
+                }
+            }
+        }
+        if tokens.next().is_some() {
+            return Err(InvalidImdsResponse::Custom(
+                "found more JSON tokens after completing parsing".into(),
+            ));
+        }
+    }
+    match code {
+        // IMDS does not appear to reply with an unset code, but documentation indicates it
+        // may be possible
+        None | Some(Cow::Borrowed("Success")) => {
```

Comment:
The `None` case can also occur if you got a JSON response with anything other than a `StartObject` token at the very beginning. This really shouldn't happen in practice, but I think maybe the first token check should have an else clause to handle this error case so that the error can be more helpful.